sklearn/neighbors/_dist_metrics.pyx

```diff
@@ -292,6 +292,14 @@ cdef class DistanceMetric:
         if self.__class__ is DistanceMetric:
             raise NotImplementedError("DistanceMetric is an abstract class")
 
+    def _validate_data(self, X):
```

Comment:
In case this is only going to be used in `cdef` classes, you can also make it a `cdef` method to avoid some call overhead. Also, if the type of `x` is always the same (kind of memoryview), type it to avoid creating a Python wrapper just for this call.

Reply:
Thanks for the comment!

> Also, if the type of x is always the same (kind of memoryview?), type it to avoid creating a Python wrapper just for this call.

It's probably a `np.ndarray[DTYPE_t, ndim=2]`, we are waiting for the stable Cython 3.0 release to migrate to memoryviews everywhere due to the need to use const fused memoryviews (https://github.com/scikit-learn/scikit-learn/issues/10624). Though here it's likely only const memoryviews, anyway it shouldn't be a performance critical part of the code, so even python overhead for non typed version shouldn't have much impact I think.

Reply:
Thanks for your comments. @rth is correct. Thanks for your answering.